File: Demos/WebView Demo for iOS/WebView Demo for iOS/WebViewController.m
//
//  WebViewController.m
//  1Password Extension Demo
//
//  Created by Dave Teare on 2014-07-19.
//  Copyright (c) 2014 AgileBits. All rights reserved.
//

#import "WebViewController.h"

#import "OnePasswordExtension.h"

@interface WebViewController() <UISearchBarDelegate, WKNavigationDelegate>

@property (weak, nonatomic) IBOutlet UIButton *onepasswordFillButton;
@property (weak, nonatomic) IBOutlet UIView *webViewContainer;
@property (weak, nonatomic) IBOutlet UISearchBar *searchBar;
@property (strong, nonatomic) WKWebView *webView;

@end

@implementation WebViewController

#pragma mark - Life Cycle

- (void)viewDidLoad {
	[super viewDidLoad];

	[self.onepasswordFillButton setHidden:![[OnePasswordExtension sharedExtension] isAppExtensionAvailable]];

	WKWebViewConfiguration *configuration = [WKWebViewConfiguration new];
	self.webView = [[WKWebView alloc] initWithFrame:self.webViewContainer.bounds configuration:configuration];
	self.webView.autoresizingMask = UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth;
	self.webView.navigationDelegate = self;
	[self.webViewContainer addSubview:self.webView];

	NSString *htmlFile = [[NSBundle mainBundle] pathForResource:@"welcome" ofType:@"html"];
	NSString* htmlString = [NSString stringWithContentsOfFile:htmlFile encoding:NSUTF8StringEncoding error:nil];
	[self.webView loadHTMLString:htmlString baseURL:nil];
}

#pragma mark - Actions

- (IBAction)fillUsing1Password:(id)sender {
	[[OnePasswordExtension sharedExtension] fillItemIntoWebView:self.webView forViewController:self sender:sender showOnlyLogins:NO completion:^(BOOL success, NSError *error) {
		if (!success) {
			NSLog(@"Failed to fill into webview: <%@>", error);
		}
	}];
}

- (IBAction)goBack:(id)sender {
	WKNavigation *navigation = [self.webView goBack];

	if (!navigation) {
		NSString *htmlFile = [[NSBundle mainBundle] pathForResource:@"welcome" ofType:@"html"];
		NSString* htmlString = [NSString stringWithContentsOfFile:htmlFile encoding:NSUTF8StringEncoding error:nil];
		[self.webView loadHTMLString:htmlString baseURL:nil];

	}
}

- (IBAction)goForward:(id)sender {
	[self.webView goForward];
}

#pragma mark - UISearchBarDelegate

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar {
	[self performSearch:searchBar.text];
}

- (void)searchBarTextDidEndEditing:(UISearchBar *)searchBar {
	[self performSearch:searchBar.text];
}

- (void)handleSearch:(UISearchBar *)searchBar {
	[self performSearch:searchBar.text];
}

- (void)searchBarCancelButtonClicked:(UISearchBar *) searchBar {
	[self performSearch:searchBar.text];
}

#pragma mark - Convenience Methods

- (void)performSearch:(NSString *)text {
	NSString *lowercaseText = [text lowercaseStringWithLocale:[NSLocale currentLocale]];
	NSURL *URL = nil;

	BOOL hasSpaces = [lowercaseText rangeOfString:@" "].location != NSNotFound;
	BOOL hasDots = [lowercaseText rangeOfString:@"."].location != NSNotFound;
	BOOL search = !hasSpaces || !hasDots;
	if (search) {
		BOOL hasScheme = [lowercaseText hasPrefix:@"http:"] || [lowercaseText hasPrefix:@"https:"];
		if (hasScheme) {
			URL = [NSURL URLWithString:lowercaseText];
		}
		else {
			URL = [NSURL URLWithString:[@"https://" stringByAppendingString:lowercaseText]];
		}
	}

	if (URL == nil) {
		NSString *escapedText = [lowercaseText stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
		NSString *googleSearch = @"http://www.google.com/search?q=";
		URL = [NSURL URLWithString:[googleSearch stringByAppendingString:escapedText]];
	}

	self.searchBar.text = [URL absoluteString];
	[self.searchBar resignFirstResponder];

	NSURLRequest *request = [[NSURLRequest alloc] initWithURL:URL];
	[self.webView loadRequest:request];
}

#pragma mark - WKNavigationDelegate

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
	self.searchBar.text = webView.URL.absoluteString;

	if ([self.searchBar.text isEqualToString:@"about:blank"]) {
		self.searchBar.text = @"";
	}
}

@end

File: Demos/WebView Demo for iOS Swift/WebView Demo for iOS Swift/WebViewController.swift
//
//  WebViewController.swift
//  WebView Demo for iOS Swift
//
//  Created by Rad Azzouz on 2015-05-14.
//  Copyright (c) 2015 AgileBits Inc. All rights reserved.
//

import UIKit

class WebViewController: UIViewController, UISearchBarDelegate, WKNavigationDelegate {

	@IBOutlet weak var onepasswordFillButton: UIButton!
	@IBOutlet weak var webViewContainer: UIView!
	@IBOutlet weak var searchBar: UISearchBar!
	@IBOutlet var webView: WKWebView!

	override func viewDidLoad() {
		super.viewDidLoad()
		self.onepasswordFillButton.isHidden = (false == OnePasswordExtension.shared().isAppExtensionAvailable())

		let configuration = WKWebViewConfiguration()
		
		self.webView = WKWebView(frame: self.webViewContainer.bounds, configuration: configuration)
		self.webView.autoresizingMask = UIViewAutoresizing(arrayLiteral: .flexibleHeight, .flexibleWidth)
		self.webView.navigationDelegate = self
		self.webViewContainer.addSubview(self.webView)

		let htmlFilePath = Bundle.main.path(forResource: "welcome", ofType: "html")
		var htmlString : String!
		do {
			htmlString = try String(contentsOfFile: htmlFilePath!, encoding: String.Encoding.utf8)
		}
		catch {
			print("Failed to obtain the html string from file \(String(describing: htmlFilePath)) with error: <\(String(describing: error))>")
		}

		self.webView.loadHTMLString(htmlString, baseURL: URL(string: "https://agilebits.com"))
	}

	@IBAction func fillUsing1Password(_ sender: AnyObject) -> Void {
		OnePasswordExtension.shared().fillItem(intoWebView: self.webView, for: self, sender: sender, showOnlyLogins: false) { (success, error) -> Void in
			if success == false {
				print("Failed to fill into webview: <\(String(describing: error))>")
			}
		}
	}

	@IBAction func goBack(_ sender: AnyObject) -> Void {
		let navigation = self.webView.goBack()

		if navigation == nil {
			let htmlFilePath = Bundle.main.path(forResource: "welcome", ofType: "html")
			var htmlString : String!
			do {
				htmlString = try String(contentsOfFile: htmlFilePath!, encoding: String.Encoding.utf8)
			}
			catch {
				print("Failed to obtain the html string from file \(String(describing: htmlFilePath)) with error: <\(String(describing: error))>")
			}

			self.webView.loadHTMLString(htmlString, baseURL: URL(string: "https://agilebits.com"))
		}
	}
	@IBAction func goForward(_ sender: AnyObject) -> Void {
		self.webView.goForward()
	}

	// UISearchBarDelegate
	func searchBarSearchButtonClicked(_ searchBar: UISearchBar) {
		self.performSearch(text: searchBar.text)
	}

	func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
		self.performSearch(text: searchBar.text)
	}

	func handleSearch(searchBar: UISearchBar) {
		self.performSearch(text: searchBar.text)
	}

	func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
		self.performSearch(text: searchBar.text)
	}

	// Convenience
	func performSearch(text: String!) {
		let lowercaseText = text.lowercased(with: NSLocale.current)
		var url: URL?

		let hasSpaces = lowercaseText.range(of: " ") != nil
		let hasDots = lowercaseText.range(of: ".") != nil

		let search: Bool = !hasSpaces || !hasDots
		if (search) {
			let hasScheme = lowercaseText.hasPrefix("http:") || lowercaseText.hasPrefix("https:")
			if (hasScheme) {
				url = URL(string: lowercaseText)
			}
			else {
				url = URL(string: "https://" + lowercaseText)
			}
		}

		if (url == nil) {
			let urlComponents = NSURLComponents()
			urlComponents.scheme = "https"
			urlComponents.host = "www.google.com"
			urlComponents.path = "/search"
			
			let queryItem = URLQueryItem(name: "q", value: text)
			urlComponents.queryItems = [queryItem]
			
			url = urlComponents.url
		}

		self.searchBar.text = url?.absoluteString
		self.searchBar.resignFirstResponder()

		let request = URLRequest(url: url!)
		self.webView.load(request)
	}
	
	// WKNavigationDelegate
	func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
		self.searchBar.text = webView.url?.absoluteString

		if self.searchBar.text == "about:blank" {
			self.searchBar.text = ""
		}
	}
}

